### PR TITLE
[kjobctl] Add --time flag on Slurm mode.

### LINKED ISSUE #3332: [kjobctl] Support slurm-like -t option based on max-execution-time
**What would you like to be added**:

Support slurm-like -t option based on max-execution time.

This is follow up to https://github.com/kubernetes-sigs/kueue/issues/3125.

**Why is this needed**:

To support slurm-like timeout in kjobctl.

### PR DESCRIPTION
**Pull request by mbobrovskyi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Add --time flag on kjobctl Slurm mode.

#### Which issue(s) this PR fixes:

Fixes #3332

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: cmd/experimental/kjobctl/go.mod [hunk 1]
```diff
@@ -18,7 +18,7 @@ require (
 	k8s.io/kubectl v0.31.2
 	k8s.io/utils v0.0.0-20240902221715-702e33fdd3c3
 	sigs.k8s.io/controller-runtime v0.19.0
-	sigs.k8s.io/kueue v0.8.1
+	sigs.k8s.io/kueue v0.9.0-rc.1.0.20241025140252-9bd0fccfb464
```
Comment (mbobrovskyi) [inline cmd/experimental/kjobctl/go.mod:21]: /hold for release

File: cmd/experimental/kjobctl/apis/v1alpha1/application_profile_types.go [hunk 2]
```diff
@@ -93,6 +94,7 @@ type TemplateReference string
 // +kubebuilder:validation:XValidation:rule="!has(self.requiredFlags) || !('ntasks' in self.requiredFlags) || self.name == 'Slurm'", message="ntasks flag can be used only on Slurm mode"
 // +kubebuilder:validation:XValidation:rule="!has(self.requiredFlags) || !('output' in self.requiredFlags) || self.name == 'Slurm'", message="output flag can be used only on Slurm mode"
 // +kubebuilder:validation:XValidation:rule="!has(self.requiredFlags) || !('partition' in self.requiredFlags) || self.name == 'Slurm'", message="partition flag can be used only on Slurm mode"
+// +kubebuilder:validation:XValidation:rule="!has(self.requiredFlags) || !('time' in self.requiredFlags) || self.name == 'Slurm'", message="time flag can be used only on Slurm mode"
```
Comment (mszadkow) [inline cmd/experimental/kjobctl/apis/v1alpha1/application_profile_types.go:97]: Is this flag slurm only for real or just slurp to begins with?
Will other modes also get it in the next step?

Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/apis/v1alpha1/application_profile_types.go:97]: I think we can add to other modes in follow-ups. But I'm not sure that it should be named as "time".

File: cmd/experimental/kjobctl/pkg/parser/slurm_test.go [hunk 3]
```diff
@@ -94,6 +94,7 @@ sleep 10`,
 #SBATCH --time=1:00:00
 #SBATCH --mail-type=END
 #SBATCH --mail-user=your@email.com
+#SBATCH --unknown=unknown
```
Comment (mszadkow) [inline cmd/experimental/kjobctl/pkg/parser/slurm_test.go:97]: what is the unknown?

Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/slurm_test.go:97]: In this test, the parser should skip unknown flags and not return an error. In the original version, we used `--time` as an unknown flag, but now it's recognized.

**Comment by mbobrovskyi** [comment]
> Is there a reason why we're not handling infinity like in slurm? so keywords: "infinite", "unlimited", "Infinite", "Unlimited", "INFINITE", "UNLIMITED", and "-1"?

Good point. I missed that. Thanks!

**Review by mwysokin — CHANGES_REQUESTED** [review]
Sorry, I had a couple of comments stuck in `PENDING` 🤦 

There's one bug which I believe prevents the PR from being merged.

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"k8s.io/utils/ptr"
+)
+
+func TestParseTimeLimit(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:27]: If it "0" we should run INFINITE too, correct?

File: cmd/experimental/kjobctl/pkg/parser/time_limit.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	if digit == 0 {
+		return false
+	}
+
+	if dash > 1 || colon > 2 {
+		return false
+	}
+
+	if dash > 0 {
+		if colon == 1 && digit > 3 {
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/time_limit.go]: Fixed. Thanks!

File: cmd/experimental/kjobctl/pkg/cmd/printcrds/embed/manifest.gz
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/cmd/printcrds/embed/manifest.gz:1]: This is embedded CRDs introduced on https://github.com/kubernetes-sigs/kueue/pull/2695.

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:27]: > A time limit of zero requests that no time limit be imposed.

https://slurm.schedmd.com/sbatch.html#OPT_time. Looks like yes.

File: cmd/experimental/kjobctl/docs/run_slurm.md [hunk 6]
```diff
@@ -38,6 +38,7 @@ Kjob provides support for executing Slurm scripts by offering several options th
 | -n, --ntasks        | Specifies the number of identical containers inside of a pod, usually 1. |
 | -o, --output        | Specifies where to redirect the standard output stream of a task. If not passed, it proceeds to stdout and is available via `kubectl logs`. |
 | --partition         | Specifies the local queue name. See [Local Queue](https://kueue.sigs.k8s.io/docs/concepts/local_queue/) for more information. |
+| -t, --time          | Set a limit on the total run time of the job. For more details, see https://slurm.schedmd.com/sbatch.html#OPT_time. |
```
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/docs/run_slurm.md]: Done

File: cmd/experimental/kjobctl/pkg/cmd/printcrds/embed/manifest.gz
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/cmd/printcrds/embed/manifest.gz:1]: After update CRDs it's automatically updating. So yes, we need to update it.

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:27]: Done

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		},
+		"UNLIMITED": {
+			val: "UNLIMITED",
+		},
+		"infinite": {
+			val: "infinite",
+		},
+		"unlimited": {
+			val: "unlimited",
+		},
```
Comment (mwysokin) [inline cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:48]: please, excuse my ignorance, I'm probably missing something but don't we need `want`s here?

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		"single digit minimal values 0": {
+			val: "0",
+		},
+		"single digit minimal values 0:0:0": {
+			val: "0:0:0",
+		},
+		"single digit minimal values 0-0:0:0": {
+			val: "0-0:0:0",
+		},
```
Comment (mwysokin) [inline cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:69]: And also here?

**Review by mwysokin — APPROVED** [review]
Overal LGTM 🖖 

I have 2 clarifying questions. 

Thank you so much for promptly implementing the changes requested 🙏

File: cmd/experimental/kjobctl/pkg/parser/time_limit_test.go (hunk 8 above)
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/pkg/parser/time_limit_test.go:69]: Here is the same situation.

**Review by mimowo — COMMENTED** [review]
/lgtm
/approve
Thanks!

File: cmd/experimental/kjobctl/go.mod (hunk 1 above)
Comment (mimowo) [reply cmd/experimental/kjobctl/go.mod:21]: I'm actually inclined to merge it as is, and bump in a separate PR once 0.9.0 is released.

Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/go.mod:21]: /unhold

Comment (mimowo) [reply cmd/experimental/kjobctl/go.mod:21]: /hold cancel
Let me open an issue to remember about bumping it: https://github.com/kubernetes-sigs/kueue/issues/3385

File: cmd/experimental/kjobctl/apis/v1alpha1/application_profile_types.go (hunk 2 above)
Comment (mbobrovskyi) [reply cmd/experimental/kjobctl/apis/v1alpha1/application_profile_types.go:97]: Created an issue https://github.com/kubernetes-sigs/kueue/issues/3387.

**Pull request merged**